Implement union peripherals with clustering (TCA single mode / split mode on attiny's)

hammerlink opened the pull request:
This PR addresses part of issue #4, specifically adding support for union types. 

The TCA0 modes were already implemented at some point, but the functionality was lost in the commit history. I reintroduced them to peripherals with both modes clustered.

I noticed that the svd2rust crate allows peripherals to use **clustered** registers instead of direct registers. This feature allows a good handling for the TCA Peripheral as it has 2 different modes, Single or Split. 

Example solution
```rust

        dp.tca0.tca_single().ctrla().write(|w| w.clksel().div1());
        dp.tca0.tca_split().ctrla().write(|w| w.enable().set_bit());
```

Clustering example in svd
```xml
    <peripheral>
      <name>TCA0</name>
      <description>16-bit Timer/Counter Type A</description>
      <baseAddress>0x00000A00</baseAddress>
      <registers>
        <!-- default registers come here <register>...</register> -->
        
        <!--clusterered registers -->
        <cluster>
          <name>TCA_SPLIT</name>
          <description>16-bit Timer/Counter Type A - Split Mode</description>
          <addressOffset>0x0</addressOffset>

          <!-- clustered registers <register>...</register> -->
        </cluster>
      </registers>
    </peripheral>

```

Example ATDF:
```xml
        <!-- Sub register group referenced by the union, contains the actual registers  -->
        <register-group caption="16-bit Timer/Counter Type A - Single Mode"
                         name="TCA_SINGLE"
                         size="0x40">
            <register caption="Compare 0" name="CMP0" offset="0x28" rw="RW" size="2"/>
            <!-- multiple <register>...</register> -->
         </register-group>

        <!-- Module union register group, targets the sub register groups -->
        <register-group caption="16-bit Timer/Counter Type A" class="union" name="TCA" size="0x40" union-tag="TCA.SINGLE.CTRLD.SPLITM">
            <register-group name="SINGLE" name-in-module="TCA_SINGLE" offset="0" union-tag-value="0"/>
            <register-group name="SPLIT" name-in-module="TCA_SPLIT" offset="0" union-tag-value="1"/>
        </register-group>
```


Testing: Tests passed locally.

Please review and let me know if additional changes or tests are needed!

## notes
This only partially fixes the #4, as there are sub register-groups without the module register group being a union class (f.i. avr64du28 - "USB" peripheral). This was out of scope for my changes. This would need further analyses of how to include the registers of these sub groups. I already prepared the atdf peripherals to include these references, but without further implementation. It looks like these sub register-groups can be recursive as well.

hammerlink committed:
implement union peripherals with clustering (TCA single mode / split mode on attiny's)

hammerlink committed:
run cargo fmt

Rahix reviewed (changes requested):
Hi, thanks a lot for working on this!  I think the approach looks very reasonable.  Although I could not find notes about using SVD clusters for overlapping regions, if `svd2rust` can handle it, it shall be fine for us. The resulting API from `svd2rust` makes sense to me at least.

I left some detailed review comments below, but generally I think there is still a bit of chaos in the data-structures representing the register groupings. Specifically, I do not like the way `RegisterGroupHeader` is structured, for a few reasons:

- The name `Header` does not really make sense when the actual registers are stored inside as well.
- You allow storing a lot of "unused" data, in various ways.  This is a recipe for bugs because it means some data from the `Chip` structure is silently ignored during SVD generation.
  * The `Peripheral` contains a `BTreeMap` of `RegisterGroupHeader` but you only ever access one of them directly via `module_register_group_header()`.  The other "headers" are then referenced indirectly via `get_union_register_group_headers()`.  To me, these should be subelements then.
  * You have a lot of code gated on `if p.is_union() { ... } else { ... }` which is a pointer to me that there should really be an enum separating these two options instead.  Right now, you can store registers alongside register groups but only one of them will be used.
  * Related, only the `union` class is currently dealt with.  I don't think we should allow storing other classes then, when they cannot be handled properly.
  * There is a field for ungrouped registers inside `RegisterGroupHeader`, which seems to me like it could just be the `registers` from `Peripheral` instead?

I feel, it would be more reasonable to do it as follows.  This should only represent data that's actually used and prevents bugs from accessing things that aren't actually in use.


```rust
pub struct Peripheral {
    // ...

    // this is just the top-level group we find in the peripheral instance. From
    // my findings, there should only ever be one `<register-group /> inside each
    // peripheral instance.  I think we should definitely error out if we ever
    // find more, because this will need to be fixed then.
    pub registers: RegisterGroup,
}

pub struct RegisterGroup {
    pub name: String,
    pub description: Option<String>,
    pub offset: usize,

    pub registers: BTreeMap<String, Register>,
    pub subgroups: Vec<RegisterSubgroup>,
}

pub enum RegisterSubgroup {
    Union(RegisterGroupUnion),
    // this would be for the remainder of issue #4, subgroups for creating
    // structure.  We should properly detect arrays so svd2rust can built the
    // correct accessors.  But this should be added in a later PR I think.
    Single(RegisterGroup),
    Array(RegisterGroupArray),
}

pub struct RegisterGroupUnion {
    pub name: String,
    pub description: Option<String>,
    pub offset: usize,

    pub variants: BTreeMap<String, RegisterGroup>,
}
```

The parsing then needs to be adjusted accordingly.  I think it makes sense to create a new `src/atdf/register_group.rs` module module for the register group parsing.

---

I pushed a change to `main` to add the `ATxmega128A1` as another regression testcase.  Please rebase to pull it in for your changeset as well.

hammerlink commented on src/svd/peripheral.rs:
```diff
@@ -51,11 +62,42 @@ pub fn generate(p: &chip::Peripheral) -> crate::Result<svd_rs::Peripheral> {
-        .collect::<Result<Vec<_>, _>>()?;
+    let registers: Vec<svd_rs::RegisterCluster> = match p.is_union() {
+        true => {
+            println!("cargo:warning=svd generate union style {}", p.name);
```
Will be dropped in my next commit (was some residue from developing)

hammerlink committed:
Merge branch 'Rahix:main' into main

Rahix commented on src/chip.rs:
```diff
@@ -52,6 +94,33 @@ pub enum ValueRestriction {
+#[derive(Debug, Clone)]
+pub struct RegisterGroupHeader {
+    pub name: String,
+    pub class: Option<String>,
```
> There are supposed to be others like (struct, bitfield, array).

Out of curiosity, where did you get this info?

> The newer/current ATDF files coming from the Atmel site seem to use another format.

Are all ATDFs updated for this newer schema?  Or just some newer MCUs?

hammerlink replied:
In avr-device the following list would be updated because of the unions
![image](https://github.com/user-attachments/assets/c4b89041-8676-4e9f-8804-329339904ac5)

hammerlink committed:
Merge branch 'Rahix:main' into main

hammerlink committed:
Implement register group restructure + recursive implementation (#1)

* WIP - improve the register group structure and update the parsing
accordingly

* fix the dynamic address calculation

* remove unused base_address fn

* rename parse_list args & run cargo fmt

hammerlink commented:
Thanks for the thorough review, it is much appreciated!

## Implementation Updates

- I've restructured the register group handling according to your suggestions, which provides a much cleaner representation of the data and eliminates the potential for unused/ignored data issues.

- While implementing these changes, I encountered cases in actual ATDF files (such as `avr64du28.atdf`) that demonstrate nested register groups requiring recursive handling:

```xml
<register-group caption="USB Device Controller EP_TABLE" name="USB_EP_TABLE" size="0x122">
   <!-- containing multiple registers -->
   <register-group caption="USB Device Controller EP"
                   count="0x10"
                   name="EP"
                   name-in-module="USB_EP_PAIR"
                   offset="0x20"
                   size="0x100"/>
</register-group>
<register-group caption="USB Device Controller EP_PAIR" name="USB_EP_PAIR" size="0x10">
   <register-group caption="USB Device Controller OUT"
                   name="OUT"
                   name-in-module="USB_EP"
                   offset="0x00"
                   size="0x08"/>
   <register-group caption="USB Device Controller IN"
                   name="IN"
                   name-in-module="USB_EP"
                   offset="0x08"
                   size="0x08"/>
</register-group>
<register-group caption="USB Device Controller EP" name="USB_EP" size="0x8">
   <!-- containing multiple registers -->
</register-group>
```

- This led to a full recursive implementation that handles both the union and non-union cases, unintentionally implementing issue #4 completely.

## Address & Offset Handling

I noticed some complexities with the current address/offset calculation approach:

- The original implementation stored absolute addresses on registers and calculated base addresses afterward
- This approach becomes problematic with nested register groups
- The calculated base addresses often differ from what's defined in the ATDF instance

I've modified this to more closely align with how ATDF and SVD both use relative offsets between parents and children, which simplifies the nested register-group handling. These leads to small changes in the regression tests & the generated code. But the functionality should be similar. Or should I refactor this to keep everything as similar as possible?

## Question About Address Blocks

I found that the address block in the SVD peripheral doesn't appear to impact the svd2rust generation (tested in [this commit](https://github.com/hammerlink/atdf2svd/commit/940f85fb2f7cecde7988f81640833787b864d799)).
I have added a justfile to that branch if you want to see how I generated these files.

If the primary purpose of atdf2svd is to generate code for avr-device, would it make sense to either:
- Remove the address block generation entirely
- Replace it with a simpler version that combines the base address with the size of the main register group?

Because right now the address block implementation is not entirely similar. And I was wondering if it is even necessary that I refactor it to match the existing behavior.

## Additional Changes

I've noticed some additional peripheral registers appearing in the output due to these changes. 
I'm reviewing these additions but I wanted your feedback before I proceed with any further changes.

I've rebased the branch to include the ATxmega128A1 regression test as requested.

hammerlink commented on src/svd/chip.rs:
```diff
@@ -45,6 +45,12 @@ pub fn generate(c: &chip::Chip) -> crate::Result<svd_rs::Device> {
+    }
     let regs = !peripheral.registers.is_empty();
     if !regs {
         log::warn!("No registers found for peripheral {}", peripheral.name);
```
code has been removed

hammerlink commented on src/svd/chip.rs:
```diff
@@ -45,6 +45,12 @@ pub fn generate(c: &chip::Chip) -> crate::Result<svd_rs::Device> {
+        return peripheral
+            .get_union_register_group_headers()
+            .iter()
+            .any(|(header, _)| !header.registers.values().len() > 0)
```
code has been removed

hammerlink commented on src/chip.rs:
```diff
@@ -52,6 +94,33 @@ pub enum ValueRestriction {
+    pub name: String,
+    pub name_in_module: Option<String>,
+    pub description: Option<String>,
+    pub size: Option<usize>,
+    pub offset: Option<usize>,
+    pub count: Option<usize>,
```
Good point, I have removed the option from offset, and defaulted it to 0.
- `size` can be useful for defining address blocks in the svd peripheral, but removed it for now
- `count` I am not entirely sure about, I have removed it

hammerlink commented on src/chip.rs:
```diff
@@ -18,15 +18,57 @@ pub struct Chip {
+                .map(|r| r.address)
+                .min();
+        }
         self.registers.values().map(|r| r.address).min()
```
code was removed

hammerlink commented on src/atdf/register.rs:
```diff
@@ -84,3 +84,150 @@ pub fn parse(
+        register_group_header_el.iter_children_with_name("register", Some("register-group"))
+    {
+        registers.push(atdf::register::parse(register, offset, value_groups)?);
+    }
```
I merged the 2 statements into 1

hammerlink commented on src/atdf/register.rs:
```diff
@@ -84,3 +84,150 @@ pub fn parse(
+        let description = register_group_header_el
+            .attributes
+            .get("caption")
+            .and_then(|d| if !d.is_empty() { Some(d) } else { None })
```
applied in register_group.rs

hammerlink commented on src/atdf/register.rs:
```diff
@@ -84,3 +84,150 @@ pub fn parse(
+        let class = register_group_header_el
+            .attributes
+            .get("class")
+            .and_then(|d| if !d.is_empty() { Some(d) } else { None })
```
removed class from the register group struct

hammerlink commented on src/atdf/register.rs:
```diff
@@ -84,3 +84,150 @@ pub fn parse(
+                } else {
+                    None
+                }
+            });
```
removed the size code

hammerlink commented on src/atdf/register.rs:
```diff
@@ -84,3 +84,150 @@ pub fn parse(
+        let description = register_group_item_el
+            .attributes
+            .get("caption")
+            .and_then(|d| if !d.is_empty() { Some(d) } else { None })
```
removed description from the reference

hammerlink commented on src/atdf/register.rs:
```diff
@@ -84,3 +84,150 @@ pub fn parse(
+        let name_in_module = register_group_item_el
+            .attributes
+            .get("name-in-module")
+            .and_then(|d| if !d.is_empty() { Some(d) } else { None })
```
applied in register_group.rs reference struct

hammerlink commented on src/atdf/register.rs:
```diff
@@ -84,3 +84,150 @@ pub fn parse(
+
+        let size = register_group_item_el.attributes.get("size").and_then(|d| {
+            if !d.is_empty() {
+                util::parse_int(d).ok()
```
the size code was removed

hammerlink commented on src/atdf/register.rs:
```diff
@@ -84,3 +84,150 @@ pub fn parse(
+                } else {
+                    None
+                }
+            });
```
the count code was removed

hammerlink commented on src/svd/peripheral.rs:
```diff
@@ -51,11 +62,42 @@ pub fn generate(p: &chip::Peripheral) -> crate::Result<svd_rs::Peripheral> {
-        .map(|r| svd::register::generate(r, base).map(svd_rs::RegisterCluster::Register))
-        .collect::<Result<Vec<_>, _>>()?;
+    let registers: Vec<svd_rs::RegisterCluster> = match p.is_union() {
+        true => {
```
removed the code, but I learned that you can do an inline if {} else {}. Thanks!

hammerlink committed:
Fix regression test issues (#2)

* fix invalid address block regressions

* re-implement baselining the start address & fix address block issue

* apply PR comments

* run format

hammerlink committed:
remove register group item

hammerlink committed:
fix register-group offset docs & update

hammerlink commented:
Hi @Rahix , just checking in to see if you’ve had a chance to take a look at the updated PR. Let me know if there’s anything else I can do to move this forward. 
Thanks!

Rahix commented:
Didn't get to it yet, but your PR is on my list :)  Will ping you once I have reviewed your updated changeset!
